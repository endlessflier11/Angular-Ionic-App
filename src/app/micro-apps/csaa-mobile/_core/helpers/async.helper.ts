import { Store } from '@ngxs/store';
import { combineLatest, Subscription } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { FetchState } from '../store/states/fetch.state';

export const delay = (ms: number) => new Promise((r) => setTimeout(r, ms));

export const unsubscribeIfPresent = (...subs: Subscription[]) =>
  subs.filter((s) => !!s && typeof s.unsubscribe === 'function').forEach((s) => s.unsubscribe());

export const fetchIsLoadingFor = (store: Store, ...actions: any[]) =>
  combineLatest(actions.map((action) => store.select(FetchState.isLoading(action)))).pipe(
    map((areActionsLoading) => areActionsLoading.find(Boolean))
  );

export const fetchHasFailedFor = (store: Store, ...actions: any[]) =>
  combineLatest(actions.map((action) => store.select(FetchState.failed(action)))).pipe(
    map((areActionsLoading) => areActionsLoading.find(Boolean)),
    filter((failed) => failed)
  );
