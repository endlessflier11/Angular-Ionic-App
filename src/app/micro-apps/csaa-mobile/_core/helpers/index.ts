export * from './router.helpers';
export * from './misc.helpers';
export * from './async.helper';
export * from './error.helper';
export * from './string.helper';
