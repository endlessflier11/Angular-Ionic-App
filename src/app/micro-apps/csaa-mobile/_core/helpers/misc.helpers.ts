import { UniHttpErrorResponse } from '../services/http/uni-http.model';

export const noop = () => {};

export function isHttpError(error) {
  return error instanceof UniHttpErrorResponse;
}

export function getErrorStatus(error: any) {
  return isHttpError(error) ? (error as UniHttpErrorResponse).status : null;
}

const getDefault = (obj, key) => obj[key] || `(no ${key})`;
const stringifyErrorObject = (error) => {
  let entireErrorObject = '';
  try {
    entireErrorObject = JSON.stringify(error);
  } catch (e) {
    entireErrorObject = `unable to stringify error object due to ${e?.message}`;
  }
  return entireErrorObject;
};

function getHttpErrorDetails(error) {
  let errorObject = {} as any;
  try {
    if (typeof error.error === 'object') {
      // sometimes there is a pretty object
      errorObject = { ...error.error };
    } else {
      // trying to parse backend error JSON object
      errorObject = JSON.parse(error.error);
    }
  } catch {
    errorObject = { httpCode: error.status, httpMessage: error.message };
    if (typeof error.error === 'string') {
      try {
        error.message = error.error;
      } catch {
        // ignore
      }
    }
    errorObject.timeStamp = error.headers?.date;
    if (error.headers && error.headers['content-type']) {
      errorObject.moreInformation = `content-type: ${
        error.headers['content-type']
      } content-length: ${getDefault(error.headers, 'content-length')}`;
    }
  }

  errorObject.message = errorObject.message || error.message || errorObject.httpMessage;
  errorObject.timestamp = errorObject.timestamp || errorObject.timeStamp;
  errorObject.httpMessage = errorObject.httpMessage || errorObject.message;
  errorObject.moreInformation = errorObject.moreInformation || errorObject.details;
  errorObject.httpCode = errorObject.httpCode || error.status;
  const errorCode = errorObject.errorCode ? `Error code: ${errorObject.errorCode}` : '';
  const httpCode = getDefault(errorObject, 'httpCode');
  const httpMessage = getDefault(errorObject, 'httpMessage');
  const description = `${httpCode} ${httpMessage} ${errorCode}`;
  const requestUrl = getDefault(error, 'requestUrl');
  const requestOptions = error?.requestOptions || {};
  const requestMethod = getDefault(requestOptions, 'method');
  const requestResponseType = getDefault(requestOptions, 'responseType');

  return `${error.name || 'HTTP Error'} ${description}
- Request: ${requestMethod} ${requestUrl}
- Response Type: ${requestResponseType}
- URL: ${getDefault(error, 'url')}
- More Information: ${getDefault(errorObject, 'moreInformation')}
- Timestamp: ${getDefault(errorObject, 'timestamp')}
- Message: ${getDefault(errorObject, 'message')}
- Error Object: ${stringifyErrorObject(error)}`;
}

export const getErrorReason = (error) => {
  try {
    if (isHttpError(error)) {
      return getHttpErrorDetails(error);
    }
    return `${getDefault(error, 'message')}
- Stack: ${getDefault(error, 'stack')}
- Error Object: ${stringifyErrorObject(error)}`;
  } catch (e) {
    return `Unable to get the reason of the error due to: ${e}. Actual error: ${(
      error || 'undefined?'
    ).toString()}`;
  }
};
