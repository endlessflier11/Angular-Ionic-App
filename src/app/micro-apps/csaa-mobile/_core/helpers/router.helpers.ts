import { compile } from 'path-to-regexp';

import { namedRoutes } from '../../route-names.config';
import { NamedRoute, NamedRouteConfig, RouteParam } from '../interfaces';

const normalizedRoutes: NamedRoute[] = ((routeConfig: NamedRouteConfig[]): NamedRoute[] => {
  const asConfigItem = (parent: string, routeItem: NamedRouteConfig): NamedRoute => ({
    name: routeItem.name,
    path: parent + '/' + routeItem.path,
    relativePath: routeItem.path,
  });
  const prepareChildren = (
    parent: string,
    children: NamedRouteConfig[],
    acc: NamedRoute[] = []
  ) => {
    // eslint-disable-next-line @typescript-eslint/prefer-for-of
    for (let i = 0; i < children.length; i++) {
      const currentRoute: NamedRouteConfig = children[i];
      acc.push(asConfigItem(parent, currentRoute));
      const routeChildren = currentRoute.children;
      if (routeChildren) {
        const currentPath = currentRoute.path ? parent + '/' + currentRoute.path : parent;
        prepareChildren(currentPath, routeChildren, acc);
      }
    }
    return acc;
  };
  return prepareChildren('', routeConfig);
})(namedRoutes);

export class RouterHelpers {
  static getRouteRelativePath = (name: string, fallback?: string): string => {
    const namedRoute = normalizedRoutes.find((r) => r.name === name);
    if (namedRoute) {
      return namedRoute.relativePath;
    } else {
      if (fallback) {return fallback;}
      const e = new Error(`Route [${name}] not defined`);
      console.error(e);
      throw e;
    }
  };

  static getRoutePath = (name: string, params?: RouteParam): string => {
    const namedRoute = normalizedRoutes.find((r) => r.name === name);
    if (namedRoute) {
      const path = compile(namedRoute.path)(params);
      return path === '/' ? '' : path;
    } else {
      const e = new Error(`Route [${name}] not defined`);
      console.error(e);
      throw e;
    }
  };
}
