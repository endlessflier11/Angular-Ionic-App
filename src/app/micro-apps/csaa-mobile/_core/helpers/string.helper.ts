export class StringHelper {
  public static ucFirst(s: string): string {
    return s.toLowerCase().replace(/^\w/, s.substr(0, 1).toUpperCase());
  }
}
