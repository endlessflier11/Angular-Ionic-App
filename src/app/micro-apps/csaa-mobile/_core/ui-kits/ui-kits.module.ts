import { NgModule } from '@angular/core';
import { CommonModule, DatePipe } from '@angular/common';
import { ReactiveFormsModule } from '@angular/forms';
import { IonicModule } from '@ionic/angular';
import { ThemeIconComponent } from './theme-icon/theme-icon.component';
import { WorkingHoursComponent } from './working-hours/working-hours.component';
import { MyPolicyCardComponent } from './my-policy-card/my-policy-card.component';
import { SkeletonCardComponent } from './skeleton-card/skeleton-card.component';
import { BackButtonComponent } from './back-button/back-button.component';
import { PhonePipe } from './pipes/phone-number.pipe';
import { CsaaAutoIdCardComponent } from './csaa-auto-id-card/csaa-auto-id-card.component';
import { CsaaLoadingComponent } from './csaa-loading/csaa-loading.component';
import { ContentRendererComponent } from './content-renderer/content-renderer.component';
import { NationalCustomAlertModalComponent } from './national-custom-alert/national-custom-alert-modal';
import { SafeDataPipe } from './pipes/safe-data/safe-data.pipe';
import { DocumentActionModalComponent } from './document-action-modal/document-action-modal.component';
import { CsaaPolicySubtitlesComponent } from './csaa-policy-subtitles/csaa-policy-subtitles.component';
import { PaperlessTermsComponent } from './paperless-terms/paperless-terms.component';
import { PaperlessTermsModalComponent } from './paperless-terms-modal/paperless-terms-modal.component';
import { AppleWalletButtonComponent } from './apple-wallet-button/apple-wallet-button.component';

const UI_COMPONENTS = [
  ThemeIconComponent,
  WorkingHoursComponent,
  MyPolicyCardComponent,
  SkeletonCardComponent,
  BackButtonComponent,
  CsaaAutoIdCardComponent,
  CsaaLoadingComponent,
  ContentRendererComponent,
  NationalCustomAlertModalComponent,
  DocumentActionModalComponent,
  CsaaPolicySubtitlesComponent,
  PaperlessTermsComponent,
  PaperlessTermsModalComponent,
  AppleWalletButtonComponent,
];

const UI_PIPES = [PhonePipe, SafeDataPipe];

@NgModule({
  imports: [CommonModule, ReactiveFormsModule, IonicModule],
  declarations: [...UI_COMPONENTS, ...UI_PIPES],
  exports: [...UI_COMPONENTS, ...UI_PIPES],
  providers: [DatePipe],
})
export class UiKitsModule {}
