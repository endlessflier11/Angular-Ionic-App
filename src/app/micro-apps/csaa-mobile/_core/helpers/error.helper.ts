import { Observer } from 'rxjs';

import { CsaaAppInjector } from '../../csaa-app.injector';
import { AnalyticsService } from '../services/analytics.service';
import { Category, EventName, EventType } from '../interfaces';
import { RollbarReporterService } from '../services/rollbar-reporter/rollbar-reporter.service';
import { getErrorReason } from './misc.helpers';
import { UniHttpErrorResponse } from '../services/http/uni-http.model';

export class ErrorWithReporter {
  private analyticsService: AnalyticsService;
  private reporter: RollbarReporterService;

  constructor(private readonly originalError: any) {
    Object.freeze(this.originalError);
  }

  public getOriginalError() {
    return this.originalError;
  }

  public report(): void {
    console.log('CSAA: Error Reporter: ', this.originalError, getErrorReason(this.originalError));
    if (this.originalError instanceof UniHttpErrorResponse && this.originalError.status < 500) {
      return;
    }
    this.sendReport();
  }

  public reportAlways(): void {
    console.log('CSAA: Error Reporter: ', this.originalError, getErrorReason(this.originalError));
    this.sendReport();
  }

  private sendReport(): void {
    this.getReporter().report(getErrorReason(this.originalError));
    this.getAnalyticsService().trackEvent(EventName.ERROR_NOTIFICATION, Category.global, {
      event_type: EventType.MESSAGED,
      reason: getErrorReason(this.originalError),
    });
  }

  private getReporter(): RollbarReporterService {
    if (!this.reporter) {
      this.reporter = CsaaAppInjector.injector.get(RollbarReporterService);
    }
    return this.reporter;
  }

  private getAnalyticsService(): AnalyticsService {
    if (!this.analyticsService) {
      this.analyticsService = CsaaAppInjector.injector.get(AnalyticsService);
    }
    return this.analyticsService;
  }
}

export const tryAndReport = (fn, invoke?: boolean) => {
  const wrapperFn = (...args) => {
    try {
      return fn.call(fn, ...args);
    } catch (e) {
      reportError(e);
    }
  };
  if (invoke) {
    return wrapperFn();
  }
  return wrapperFn;
};

export const reportError = (e: any) => {
  new ErrorWithReporter(e).report();
};

export const forceReportError = (e: any) => {
  new ErrorWithReporter(e).reportAlways();
};

export const withErrorReporter = <T>(
  onSuccess?: (value?: T) => void,
  onError?: (e: ErrorWithReporter) => void,
  onComplete?: () => void
): Observer<T> => ({
  next: tryAndReport((value?: T) => {
    if (onSuccess) {
      onSuccess.call(onSuccess, value);
    }
  }),
  error: tryAndReport((err?: any) => {
    const errorWithReporter = new ErrorWithReporter(err);

    // If subscriber didn't provide onError fn, we'd report immediately
    // Else we'd pass the reporter with the error so they call error.report() manually
    if (!onError) {
      errorWithReporter.report();
    } else {
      onError.call(onError, errorWithReporter);
    }
  }),
  complete: tryAndReport(() => {
    if (onComplete) {
      onComplete.call(onComplete);
    }
  }),
});
