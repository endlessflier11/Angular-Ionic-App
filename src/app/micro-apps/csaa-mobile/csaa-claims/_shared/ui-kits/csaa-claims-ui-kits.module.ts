import { NgModule } from '@angular/core';
import { IonicModule } from '@ionic/angular';
import { CommonModule } from '@angular/common';
import { UiKitsModule } from '../../../_core/ui-kits/ui-kits.module';
import { AgentCardComponent } from './agent-card/agent-card.component';
import { ClaimsHistoryCardComponent } from './claims-history-card/claims-history-card.component';
import { ClaimsInfoCardComponent } from './claims-info-card/claims-info-card.component';
import { WhatToDoCardComponent } from './what-to-do-card/what-to-do-card.component';

const SHARED_COMPONENTS = [
  AgentCardComponent,
  ClaimsHistoryCardComponent,
  ClaimsInfoCardComponent,
  WhatToDoCardComponent,
];

@NgModule({
  declarations: [...SHARED_COMPONENTS],
  imports: [CommonModule, IonicModule, UiKitsModule],
  exports: [...SHARED_COMPONENTS],
})
export class CsaaClaimsUiKitsModule {}
