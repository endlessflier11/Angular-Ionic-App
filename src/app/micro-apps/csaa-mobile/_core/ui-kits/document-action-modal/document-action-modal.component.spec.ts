import { ComponentFixture, TestBed } from '@angular/core/testing';
import { IonicModule, ModalController } from '@ionic/angular';
import { ReactiveFormsModule } from '@angular/forms';
import {
  CallbackDataSelection,
  DocumentActionModalComponent,
} from './document-action-modal.component';
import { By, PageTestingModule } from '@app/testing';
import { UiKitsModule } from '../ui-kits.module';
import { delay } from '../../helpers';

describe('DocumentActionModalComponent', () => {
  let component: DocumentActionModalComponent;
  let fixture: ComponentFixture<DocumentActionModalComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [
        IonicModule,
        ReactiveFormsModule,
        PageTestingModule.withConfig({
          providesStorage: true,
          providesAnalytics: true,
          providesPlatform: true,
          providesModal: true,
        }),
        UiKitsModule,
      ],
    }).compileComponents();

    fixture = TestBed.createComponent(DocumentActionModalComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should display view document and share document', async () => {
    await delay(50);
    expect(fixture).toMatchSnapshot();
  });

  it('should display document when clicking on view document', () => {
    const modalControllerMock = TestBed.inject(ModalController) as any;
    const viewDocumentButton = fixture.debugElement.query(
      By.cssAndText('ion-item', 'View Document')
    );
    viewDocumentButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(modalControllerMock.dismiss).toHaveBeenCalledWith({
      selection: CallbackDataSelection.view,
    });
  });

  it('should share document when clicking on share document', () => {
    const modalControllerMock = TestBed.inject(ModalController) as any;
    const shareDocumentButton = fixture.debugElement.query(
      By.cssAndText('ion-item', 'Share Document')
    );
    shareDocumentButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(modalControllerMock.dismiss).toHaveBeenCalledWith({
      selection: CallbackDataSelection.share,
    });
  });

  it('should share document when clicking on save document', () => {
    const modalControllerMock = TestBed.inject(ModalController) as any;
    const shareDocumentButton = fixture.debugElement.query(
      By.cssAndText('ion-item', 'Save Document')
    );
    shareDocumentButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(modalControllerMock.dismiss).toHaveBeenCalledWith({
      selection: CallbackDataSelection.save,
    });
  });

  it('should dismiss action modal on close click', async () => {
    await delay(50);
    const closeButton = fixture.debugElement.query(By.cssAndText('ion-button', 'Close'));
    closeButton.triggerEventHandler('click', null);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
