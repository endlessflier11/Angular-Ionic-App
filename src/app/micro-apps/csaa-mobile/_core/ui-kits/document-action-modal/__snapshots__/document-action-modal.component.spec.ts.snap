// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentActionModalComponent should dismiss action modal on close click 1`] = `
<csaa-document-action-modal
  __ngContext__={[Function LRootView]}
  configService={[Function ConfigService]}
  currentTheme=""
  fileNameControl={[Function FormControl]}
  modalController={[Function AlertControllerMock]}
  platformService={[Function PlatformService]}
  subSink={[Function SubSink]}
>
  <ion-content
    class="ion-padding csaa"
    scroll-y="false"
  >
    <h2
      class="ion-text-center"
    >
      Document Options
    </h2>
    <ion-list>
      <ion-item
        detail=""
      >
        <ion-label>
          View Document
        </ion-label>
      </ion-item>
      <ion-item
        detail=""
      >
        <ion-label>
          Share Document
        </ion-label>
      </ion-item>
      <ion-item
        class="ion-margin-top csaa-save-as ion-untouched ion-pristine ion-invalid"
        lines="none"
      >
        <ion-label
          position="stacked"
        >
          Save file as
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-invalid ion-untouched ion-pristine ion-invalid"
          placeholder="File name"
          required=""
          type="text"
        />
        <span
          slot="end"
        >
          .pdf
        </span>
      </ion-item>
      <ion-item
        detail=""
        lines="none"
      >
        <ion-label>
          Save Document
        </ion-label>
      </ion-item>
    </ion-list>
  </ion-content><ion-footer
    class="ion-padding csaa"
  >
    <ion-button
      expand="full"
    >
      Close
    </ion-button>
  </ion-footer>
</csaa-document-action-modal>
`;

exports[`DocumentActionModalComponent should display view document and share document 1`] = `
<csaa-document-action-modal
  __ngContext__={[Function LRootView]}
  configService={[Function ConfigService]}
  currentTheme=""
  fileNameControl={[Function FormControl]}
  modalController={[Function AlertControllerMock]}
  platformService={[Function PlatformService]}
  subSink={[Function SubSink]}
>
  <ion-content
    class="ion-padding csaa"
    scroll-y="false"
  >
    <h2
      class="ion-text-center"
    >
      Document Options
    </h2>
    <ion-list>
      <ion-item
        detail=""
      >
        <ion-label>
          View Document
        </ion-label>
      </ion-item>
      <ion-item
        detail=""
      >
        <ion-label>
          Share Document
        </ion-label>
      </ion-item>
      <ion-item
        class="ion-margin-top csaa-save-as ion-untouched ion-pristine ion-invalid"
        lines="none"
      >
        <ion-label
          position="stacked"
        >
          Save file as
        </ion-label>
        <ion-input
          class="ng-untouched ng-pristine ng-invalid ion-untouched ion-pristine ion-invalid"
          placeholder="File name"
          required=""
          type="text"
        />
        <span
          slot="end"
        >
          .pdf
        </span>
      </ion-item>
      <ion-item
        detail=""
        lines="none"
      >
        <ion-label>
          Save Document
        </ion-label>
      </ion-item>
    </ion-list>
  </ion-content><ion-footer
    class="ion-padding csaa"
  >
    <ion-button
      expand="full"
    >
      Close
    </ion-button>
  </ion-footer>
</csaa-document-action-modal>
`;
